**INTRODUCTION OF THE DATA SET**  :- The given data set belongs to an airline carrier in which the records are showing us about,

the no.of flights are in service,

the departure and arrive timings, 

the delays in both arrival and departure,

the duration for the dealys with the Day, Month and Year.

And other details are not prvided. It seems to be data has many null values.
It means records are not available for the attributes.

**TO PERFORM THE ACTIONS ON THE DATA SET IN SPARK, WE HAVE TO INSTALL "PYSPARK"**

In [ ]:
!pip install pyspark

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


To perform different actions in pyspark we have to import some libraries

In [ ]:
from pyspark.sql import SparkSession
spark=SparkSession.builder.appName("assignment").getOrCreate()
spark.sparkContext.getConf().getAll();

In [ ]:
from pyspark import SparkContext, SparkConf
conf = SparkConf().setAppName("sample_app")
from pyspark.sql import SQLContext
import pyspark.sql.functions as f
from pyspark.sql.types import IntegerType
from pyspark.sql import SparkSession
from pyspark.sql.functions import when
from pyspark.sql.functions import udf,col
import re  # Regex Library
from pyspark.sql import functions as f

**To read the CSV file.**

In [ ]:
Airlines_data=spark.read.csv("/content/Airline_data.csv",header=True,inferSchema=True)

**Observations :** By using describe function we can actually see how the data and what the data is about. 

In [ ]:
summary = Airlines_data.describe().show()

+-------+------+-----+-----------------+-----------------+------------------+-----------------+------------------+------------------+-------------+-----------------+-------+------------------+------------------+-------+------------------+------------------+------+----+------------------+------+-------+---------+----------------+--------+------------+------------+--------+-------------+-----------------+
|summary|  Year|Month|       DayofMonth|        DayOfWeek|           DepTime|       CRSDepTime|           ArrTime|        CRSArrTime|UniqueCarrier|        FlightNum|TailNum| ActualElapsedTime|    CRSElapsedTime|AirTime|          ArrDelay|          DepDelay|Origin|Dest|          Distance|TaxiIn|TaxiOut|Cancelled|CancellationCode|Diverted|CarrierDelay|WeatherDelay|NASDelay|SecurityDelay|LateAircraftDelay|
+-------+------+-----+-----------------+-----------------+------------------+-----------------+------------------+------------------+-------------+-----------------+-------+-------------

In [ ]:
Airlines_data.count()

426


**Observations :** By using count function we can find the no.of records present in the given data set. We have total **426 records**

In [ ]:
len(Airlines_data.columns)

29

**Observations :** So, we came to know that data set has **29 attributes**

In [ ]:
Airlines_data.columns

['Year',
 'Month',
 'DayofMonth',
 'DayOfWeek',
 'DepTime',
 'CRSDepTime',
 'ArrTime',
 'CRSArrTime',
 'UniqueCarrier',
 'FlightNum',
 'TailNum',
 'ActualElapsedTime',
 'CRSElapsedTime',
 'AirTime',
 'ArrDelay',
 'DepDelay',
 'Origin',
 'Dest',
 'Distance',
 'TaxiIn',
 'TaxiOut',
 'Cancelled',
 'CancellationCode',
 'Diverted',
 'CarrierDelay',
 'WeatherDelay',
 'NASDelay',
 'SecurityDelay',
 'LateAircraftDelay']

**Observations :** If we have more columns or attributes,then we can list out the attributes and see which are present in the given data set by using above code.

1) Show a sample of 5 records from dataset.



In [ ]:
Airlines_data.show(5)

+----+-----+----------+---------+-------+----------+-------+----------+-------------+---------+-------+-----------------+--------------+-------+--------+--------+------+----+--------+------+-------+---------+----------------+--------+------------+------------+--------+-------------+-----------------+
|Year|Month|DayofMonth|DayOfWeek|DepTime|CRSDepTime|ArrTime|CRSArrTime|UniqueCarrier|FlightNum|TailNum|ActualElapsedTime|CRSElapsedTime|AirTime|ArrDelay|DepDelay|Origin|Dest|Distance|TaxiIn|TaxiOut|Cancelled|CancellationCode|Diverted|CarrierDelay|WeatherDelay|NASDelay|SecurityDelay|LateAircraftDelay|
+----+-----+----------+---------+-------+----------+-------+----------+-------------+---------+-------+-----------------+--------------+-------+--------+--------+------+----+--------+------+-------+---------+----------------+--------+------------+------------+--------+-------------+-----------------+
|1989|    1|        23|        1|   1419|      1230|   1742|      1552|           UA|      183

**Observations :** Showing the the 1st 5 rows of the dataframe as sample.

2) Read the data with data types.



In [ ]:
Airlines_data.printSchema()

root
 |-- Year: integer (nullable = true)
 |-- Month: integer (nullable = true)
 |-- DayofMonth: integer (nullable = true)
 |-- DayOfWeek: integer (nullable = true)
 |-- DepTime: integer (nullable = true)
 |-- CRSDepTime: integer (nullable = true)
 |-- ArrTime: integer (nullable = true)
 |-- CRSArrTime: integer (nullable = true)
 |-- UniqueCarrier: string (nullable = true)
 |-- FlightNum: integer (nullable = true)
 |-- TailNum: string (nullable = true)
 |-- ActualElapsedTime: integer (nullable = true)
 |-- CRSElapsedTime: integer (nullable = true)
 |-- AirTime: string (nullable = true)
 |-- ArrDelay: integer (nullable = true)
 |-- DepDelay: integer (nullable = true)
 |-- Origin: string (nullable = true)
 |-- Dest: string (nullable = true)
 |-- Distance: integer (nullable = true)
 |-- TaxiIn: string (nullable = true)
 |-- TaxiOut: string (nullable = true)
 |-- Cancelled: integer (nullable = true)
 |-- CancellationCode: string (nullable = true)
 |-- Diverted: integer (nullable = true)
 |


**Observations :** By using schema we can have a look at the Column data of which type it is. Then we can able to do our operations easily and accurately.



3) Make a new column MonthStr, Which has months in form of 01, 02, 03, ..., 12.



In [ ]:
from pyspark.sql.functions import lit
New_monthcol = Airlines_data.withColumn("Month_num", when((Airlines_data.Month == "1"), lit("01")))
New_monthcol.show(10)

+----+-----+----------+---------+-------+----------+-------+----------+-------------+---------+-------+-----------------+--------------+-------+--------+--------+------+----+--------+------+-------+---------+----------------+--------+------------+------------+--------+-------------+-----------------+---------+
|Year|Month|DayofMonth|DayOfWeek|DepTime|CRSDepTime|ArrTime|CRSArrTime|UniqueCarrier|FlightNum|TailNum|ActualElapsedTime|CRSElapsedTime|AirTime|ArrDelay|DepDelay|Origin|Dest|Distance|TaxiIn|TaxiOut|Cancelled|CancellationCode|Diverted|CarrierDelay|WeatherDelay|NASDelay|SecurityDelay|LateAircraftDelay|Month_num|
+----+-----+----------+---------+-------+----------+-------+----------+-------------+---------+-------+-----------------+--------------+-------+--------+--------+------+----+--------+------+-------+---------+----------------+--------+------------+------------+--------+-------------+-----------------+---------+
|1989|    1|        23|        1|   1419|      1230|   1742|    

**Observations** :- Here, by suing "lit" function we used to create a new column with values depend upon another column which is already in the data frame. So, with respect to the values present in the "Month" column, we created a new column values as shown above. 
For example **the value "1" in the "Month" column, we written code to create "01" as the values in the new column "Month_num"**.

4) Find the Unique carrier of flights each airline made.



In [ ]:
flightscarrier = Airlines_data.groupby('UniqueCarrier').count().sort("count").show(n=5,truncate=False)

+-------------+-----+
|UniqueCarrier|count|
+-------------+-----+
|UA           |426  |
+-------------+-----+



**Observations :** By using group by function , we came to know that we have only 1 unique carrier in the given data set. And for that unique carrier **"UA"** has **426 flights**

5) Find the mean departure delay per origination airport.



In [ ]:
from pyspark.sql.functions import sum,desc,avg
Mean_Arrival_delay = Airlines_data.groupby("Origin").agg(avg("ArrDelay").alias("Arrival_Delay"))
Mean_Arrival_delay.show(n=100,truncate=False)


+------+-------------------+
|Origin|Arrival_Delay      |
+------+-------------------+
|LIH   |0.16666666666666666|
|HNL   |14.21774193548387  |
|EWR   |9.25               |
|DEN   |20.166666666666668 |
|IAD   |12.966666666666667 |
|SFO   |11.215384615384615 |
|PHL   |6.827586206896552  |
|OGG   |16.24137931034483  |
+------+-------------------+



**Observations :** By using group by we can check the unique  Origin place of the flight and by using aggregate function we can check the average delay time of the flights from a particluar region.

6) What is the average departure delay from each airport?

In [ ]:
Mean_Departure_delay = Airlines_data.groupby("Origin").agg(avg("DepDelay").alias("Departure_Delay"))
Mean_Departure_delay.show(n=100,truncate=False)


+------+-------------------+
|Origin|Departure_Delay    |
+------+-------------------+
|LIH   |-3.7666666666666666|
|HNL   |3.217741935483871  |
|EWR   |4.958333333333333  |
|DEN   |27.6               |
|IAD   |8.9                |
|SFO   |19.646153846153847 |
|PHL   |16.137931034482758 |
|OGG   |6.0                |
+------+-------------------+



**Observations :** By using group by we can check the unique  Departure of the flight and by using aggregate function we can check the average delay time of the flights from a particluar region.